File: 54.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if not matrix or not matrix[0]: return []
        res = []
        i, di, j, dj = 0,0,0,1
        m,n = len(matrix), len(matrix[0])
        for _ in range(m*n):
            res.append(matrix[i][j])
            matrix[i][j] = ""
            if matrix[(i + di)%m][(j+dj)%n] == "":
                di, dj = dj, -di
            i += di
            j += dj
        return res

File: 54.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.size() == 0 or matrix[0].size() == 0){
            return {};
        }
        int di = 0, dj = 1;
        int i = 0, j = 0;
        vector<int> res;
        for(int k = 0; k < matrix.size() * matrix[0].size(); k++){
            res.push_back(matrix[i][j]);
            matrix[i][j] = 0;
            if(matrix[(i+di)%matrix.size()][(j+dj)%matrix[0].size()] == 0){
                int tmp = di;
                di = dj;
                dj = -tmp;
            }
            i = i + di;
            j = j + dj;
        }
        return res;
    }
};
